src/index/CompressedBlockPrefiltering.h

```diff
@@ -0,0 +1,140 @@
+//  Copyright 2024, University of Freiburg,
+//                  Chair of Algorithms and Data Structures
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de>
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "global/Id.h"
+#include "global/ValueIdComparators.h"
+#include "index/CompressedRelation.h"
+
+// The compressed block metadata (see `CompressedRelation.h`) that we use to
+// filter out the non-relevant blocks by checking their content of
+// `firstTriple_` and `lastTriple_` (`PermutedTriple`)
+using BlockMetadata = CompressedBlockMetadata;
+
+namespace prefilterExpressions {
+//______________________________________________________________________________
+/*
+`PrefilterExpression` represents a base class for the following sub-classes that
+implement the block-filtering procedure for the specific relational + logical
+operations/expressions.
+
+Remark: We do not actually evaluate the respective SPARQL Expression. We only
+pre-filter w.r.t. blocks that contain relevant data for the actual evaluation of
+those expressions to make the evaluation procedure more efficient.
+
+The block-filtering is applied with the following operations:
+Relational Expressions - `<=`, `>=`, `<`, `>`, `==` and `!=`
+Logical Operations - `and` and `or`
+*/
+
+class PrefilterExpression {
+ public:
+  // The respective metadata to the blocks is expected to be provided in
+  // a sorted order (w.r.t. the relevant column).
+  virtual std::vector<BlockMetadata> evaluate(std::vector<BlockMetadata>& input,
+                                              size_t evaluationColumn) = 0;
+  virtual ~PrefilterExpression() = default;
+};
+
+//______________________________________________________________________________
+// For the actual comparison of the relevant ValueIds from the metadata triples,
+// we use the implementations from ValueIdComparators.
+//
+// Supported comparisons are:
+//  - LessThan, LessEqual, Equal, NotEqual, GreaterEqual, GreaterThan
+using CompOp = valueIdComparators::Comparison;
+
+//______________________________________________________________________________
+// Given a PermutedTriple, retrieve the suitable Id w.r.t. a column (index).
+constexpr auto getIdFromColumnIndex =
+    [](const BlockMetadata::PermutedTriple& permutedTriple,
+       const size_t columnIndex) {
+      switch (columnIndex) {
+        case 0:
+          return permutedTriple.col0Id_;
+        case 1:
+          return permutedTriple.col1Id_;
+        case 2:
+          return permutedTriple.col2Id_;
+        default:
+          // Triple!
+          AD_FAIL();
+      }
+    };
```

Comment:
This function can be moved to the `.cpp` file, where it can be made `static`.
(Doesn't need to be part of a class, just a `static`function or lambda that is only defined in a cpp file as part of the implementation.